beacon-chain/rpc/beaconv1/state.go

```diff
@@ -1,30 +1,249 @@
 package beaconv1
 
 import (
+	"bytes"
 	"context"
-	"errors"
+	"strconv"
+	"time"
 
 	ptypes "github.com/gogo/protobuf/types"
+	"github.com/pkg/errors"
 	ethpb "github.com/prysmaticlabs/ethereumapis/eth/v1"
+	ethpb_alpha "github.com/prysmaticlabs/ethereumapis/eth/v1alpha1"
+	"github.com/prysmaticlabs/prysm/beacon-chain/db/filters"
+	"github.com/prysmaticlabs/prysm/beacon-chain/state"
+	"github.com/prysmaticlabs/prysm/shared/bytesutil"
+	"github.com/prysmaticlabs/prysm/shared/params"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
 )
 
+var stateRequestTimeout = 10 * time.Second
+
 // GetGenesis retrieves details of the chain's genesis which can be used to identify chain.
 func (bs *Server) GetGenesis(ctx context.Context, _ *ptypes.Empty) (*ethpb.GenesisResponse, error) {
-	return nil, errors.New("unimplemented")
+	genesisTime := bs.GenesisTimeFetcher.GenesisTime()
+	var defaultGenesisTime time.Time
+	var gt *ptypes.Timestamp
+	var err error
+	if genesisTime == defaultGenesisTime {
+		gt, err = ptypes.TimestampProto(time.Unix(0, 0))
+	} else {
+		gt, err = ptypes.TimestampProto(genesisTime)
+	}
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Could not convert genesis time to proto: %v", err)
+	}
+
+	genValRoot := bs.ChainInfoFetcher.GenesisValidatorRoot()
+	return &ethpb.GenesisResponse{
+		GenesisTime:           gt,
+		GenesisValidatorsRoot: genValRoot[:],
+		GenesisForkVersion:    params.BeaconConfig().GenesisForkVersion,
+	}, nil
 }
 
 // GetStateRoot calculates HashTreeRoot for state with given 'stateId'. If stateId is root, same value will be returned.
 func (bs *Server) GetStateRoot(ctx context.Context, req *ethpb.StateRequest) (*ethpb.StateRootResponse, error) {
-	return nil, errors.New("unimplemented")
+	var blk *ethpb_alpha.SignedBeaconBlock
+	var err error
+	switch string(req.StateId) {
+	case "head":
+		blk, err = bs.ChainInfoFetcher.HeadBlock(ctx)
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "Could not get head state: %v", err)
+		}
+	case "genesis":
+		blk, err = bs.BeaconDB.GenesisBlock(ctx)
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "Could not get genesis checkpoint: %v", err)
+		}
+	case "finalized":
+		finalizedCheckpoint := bs.ChainInfoFetcher.FinalizedCheckpt()
+		blk, err = bs.BeaconDB.Block(ctx, bytesutil.ToBytes32(finalizedCheckpoint.Root))
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "Could not get finalized checkpoint: %v", err)
+		}
+	case "justified":
+		justifiedCheckpoint := bs.ChainInfoFetcher.CurrentJustifiedCheckpt()
+		blk, err = bs.BeaconDB.Block(ctx, bytesutil.ToBytes32(justifiedCheckpoint.Root))
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "Could not get justified checkpoint: %v", err)
+		}
+	default:
+		if len(req.StateId) == 32 {
+			blk, err = bs.BeaconDB.Block(ctx, bytesutil.ToBytes32(req.StateId))
```

Comment:
What is req.StateId here? Is it a block root or state root? If it's a state root, this won't work.